torch/csrc/jit/export.cpp

```diff
@@ -465,32 +465,146 @@ class MethodEncoder : public EncoderBase {
+  void findTensorInBlock(const Block& block);
+
+  // recursively iterate over the whole module to collect the information of
+  // tensors and parameters
+  void collectInfo(
+      const script::Module& module,
+      const std::string& prefix);
+
+  // write the content of the tensor to the file/stream, and save the
+  // offset in the storageMap_
+  void convertAndWriteTensor(
+      const at::Tensor& tensor,
+      caffe2::TensorProto* tensor_proto);
+
+  // dump all the tensors in the tensorTable_ to a ModelDef (metadata) and
+  // the file/stream (the content), assuming all the information of the
+  // tensors has been collected. the method calls convertAndWriteTensor
+  // to dump the content of a tensor
+  void writeTensorTable(torch::ModelDef* model_def);
+
+  void convertModule(
+      const script::Module& module,
+      const std::string& name,
+      torch::ModuleDef* module_def);
+
+  void convertParameter(
+      const script::NamedParameter& param,
+      torch::ParameterDef* param_def);
+
+  void convertMethod(const script::Method& method, torch::MethodDef* method_def);
+
+  std::ofstream ofs_;
+  PyTorchStreamWriter writer_;
+  // storage_ptr => record_offset
+  std::unordered_map<const void*, uint64_t> storageMap_;
+  // tensor => param name
+  std::unordered_map<const at::Tensor*, std::string> paramMap_;
+  // tensor => tensor_id
+  std::unordered_map<const at::Tensor*, uint64_t> tensorTable_;
+  // used for generating table id for tensors
+  uint64_t nextTensorId_ = 0;
+};
+
+// MethodEncoder's methods
 MethodEncoder::MethodEncoder(
     const script::Method& method,
     std::string* torch_script,
-    std::unordered_map<const void*, uint64_t>* storage_map,
-    std::unordered_map<at::Tensor*, std::string>* parameter_map,
-    PyTorchStreamWriter* writer)
+    const ScriptModuleSerializer& serializer)
     : EncoderBase(onnx_torch::OperatorExportTypes::RAW, false) {
-  storage_dedup_map_ = storage_map;
-  parameter_map_ = parameter_map;
-  stream_writer_ = writer;
   // we already keep the tree structure in the top level module,
   // so pass "" as prefix
+  serializer_ = &serializer;
   EncodeMethod(torch_script, method, "");
 }
 
+void MethodEncoder::EncodeMethod(
+    std::string* torch_script,
+    const script::Method& method,
+    const std::string prefix) {
+  onnx::ModelProto model_proto;
+  model_proto.set_doc_string("THIS PROTO IS NOT STANDARD ONNX");
+  auto* node_proto = model_proto.mutable_graph()->add_node();
+  node_proto->set_name(prefix + method.name());
+  if (method.is_optimized()) {
+    // mark that this method was optimized
+    node_proto->set_domain("optimized");
+  }
+
+  // We store the schema string in the docstring.
+  node_proto->set_doc_string(getExportableSchemaStringForMethod(method));
+
+  // Store member_inputs of Method in input
+  for (auto &member_input : method.params()) {
+    auto it = serializer_->paramMap().find(member_input);
```

Comment:
nit: it's a nicer API if (de)serializer just exposes "lookupParam" instead of exposing the entire map.

Reply:
agree, will do it